Costume editor snapping addon

ErrorGamer2000 opened the pull request:
<!-- Which issue(s) does this pull request fix or resolve? -->

Resolves #3639

### Changes

- Addon API Changes
  - Added new shared space to `addon.tab.appendToSharedSpace` to allow addons to add controls in the paint editor.
- Addon Changes
  - Made minor visual changes to the `onion-skinning` addon.
  - Changed the `onion-skinning` settings page positioning. No longer absolute, always centered over button.
  - Changed `onion-skinning` to use the new shared space.
- New Addon
  - Added the new `paint-snap` addon:
    - Allows users to enable/disable snapping altogether, including the native snap to center.
    - Allows users to customize the snapping threshold (or the minimum distance that the selection can move to snap).
    - Allows users to enable/disable snapping to:
      - Edges of the art board / page
      - Corners of the art board / page
      - Center of the art board / page, including the axes
      - Edges of other objects on the page (Groups are considered to be a single object)
      - Centers of other objects on the page
      - Corners of other objects on the page
    - Allows users to enable/disable snapping from:
      - Center of the selection
      - Corners of the selection
      - Centers of the edges of the selection (Really just acts like the full edge, kinda)
    - Works in both vector and bitmap modes.

### Reason for changes

Paint snapping can be extremely useful in other svg editors, so why not add it to scratch.

### Tests

Can't upload video, but tested all options and snapping, as well as changes to the `onion-skinning` UI as editing was being done as well as when it was finished.

apple502j commented on addon-api/content-script/Tab.js:
```diff
@@ -357,6 +357,7 @@ export default class Tab extends Listenable {
    * assetContextMenuAfterDelete - after the delete button of asset (sprite, costume, etc)'s context menu
    * monitor - after the end of the stage monitor context menu
    *
```
Please add the description.

apple502j commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,92 @@
+  "name": "Paint snapping",
+  "description": "Snap objects in the paint editor to bounding boxes and svg nodes.",
+  "tags": ["editor", "costumeEditor", "featured"],
+  "versionAdded": "1.28.0",
```
```suggestion
  "versionAdded": "1.29.0",
```

apple502j commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,92 @@
+{
+  "$schema": "https://raw.githubusercontent.com/ScratchAddons/manifest-schema/dist/schema.json",
+  "name": "Paint snapping",
```
Needs a better name.

apple502j commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,92 @@
+{
+  "$schema": "https://raw.githubusercontent.com/ScratchAddons/manifest-schema/dist/schema.json",
+  "name": "Paint snapping",
+  "description": "Snap objects in the paint editor to bounding boxes and svg nodes.",
```
```suggestion
  "description": "Snap objects in the costume editor to bounding boxes and vector nodes.",
```

apple502j commented on addons/paint-snap/helper/math.js:
```diff
@@ -0,0 +1,175 @@
+  };
+
+  const getRandomBoolean = function () {
+    return getRandomInt(0, 2) === 1;
```
```suggestion
    return Math.random() >= 0.5;
```

apple502j commented on addons/paint-snap/lib/keyMirror.js:
```diff
@@ -0,0 +1,3 @@
+const keyMirror = (obj) => Object.fromEntries(Object.entries(obj).map(([k]) => [k, k]));
```
Don't think this is useful. I know this is a copy from vanilla, but don't need this ourselves.

apple502j commented on addons/paint-snap/state.js:
```diff
@@ -0,0 +1,74 @@
+}
+
+export function toggle(enabled) {
+  enabled ? enable() : disable();
```
Please don't.

apple502j commented on addons/paint-snap/ui.js:
```diff
@@ -0,0 +1,287 @@
+
+  const settingsPage = document.createElement("div");
+  settingsPage.className = "sa-paint-snap-settings";
+  settingsPage.dataset.compat = addonIsEnabled("onion-skinning");
```
Instead of this, make `onion-skinning` detect this addon. Do not use `getEnabledAddons` for this. (I'll probably make this crash if this is called from other addons at some point.)

ErrorGamer2000 replied:
I dont even need this, will remove. Sorry. Also will remove `window.paper = paper`

apple502j reviewed (changes requested):
Also, please fix the code spaghetti. One big file, and a small file for splittable parts of code. Please don't use module-scoped variable except for addon compatibility.

Norbiros commented:
It is great addon, but I have couple of suggestions:
1. Why you have options in addon's settings and in own UI. I think that we should remove addon's settings and stay with own ui (you can save selected options in localStorage (something like in `messages filter` addon)).
2. Why you have so much options enabled by default. We should add default configuration for example:
![Screenshot from 2022-09-01 20-55-02](https://user-images.githubusercontent.com/89837724/187990625-819d30e8-bf5a-44dc-ad5f-b460c838f021.png)
3. Maybe you can make icons more grey?
![Screenshot from 2022-09-01 21-00-05](https://user-images.githubusercontent.com/89837724/187991460-80bb5378-0d20-452d-8eeb-48273b569bd1.png)
4. I can type values like: -5, 1000 or letters to `snapping distance` input.
5. I have console error: `Uncaught TypeError: hideGuides is not a function updateSelectTool.js:545:5`

ErrorGamer2000 commented:
> It is great addon, but I have couple of suggestions:
> 
> 1. Why you have options in addon's settings and in own UI. I think that we should remove addon's settings and stay with own ui (you can save selected options in localStorage (something like in `messages filter` addon)).
> 2. Why you have so much options enabled by default. We should add default configuration for example:
> 3. Maybe you can make icons more grey?
> 4. I can type values like: -5, 1000 or letters to `snapping distance` input.
> 5. I have console error: `Uncaught TypeError: hideGuides is not a function updateSelectTool.js:545:5`
> 6. Can you add dynamic disable/enable.

@Norbiros Thanks for bringing this to my attention.

1. This is a better idea, working on it now.
2. I just set them all to b enabled by default. That actually looks like a good default.
3. I don't know why your zoom controls are like that... 
![image](https://user-images.githubusercontent.com/70162741/188012753-223f9dbe-7633-4957-80b4-5ad7fb965c47.png)
I actually made them the same exact color as the `onion-skinning` icons.
4. Fixed it.
5. Also fixed. Forgot to remove some code.
6. Forgot to add it to the manifest. On it.

WorldLanguages commented:
I think a toggle would be better than "on/off" because those 2 words, when translated into other languages, can become pretty long.

WorldLanguages commented:
> Maybe you can make icons more grey?

@Norbiros Are you using the vanilla colors preset of the editor dark mode addon? Looks like ErrorGamer can't reproduce that.

ErrorGamer2000 commented:
> I think a toggle would be better than "on/off" because those 2 words, when translated into other languages, can become pretty long.

I think I'll do a checkbox instead. Probably would be better than a toggle because I don't think that a toggle will match the scratch UI. I was just doing the text toggles because I wanted to match onion skinning.

mxmou commented:
> I don't know why your zoom controls are like that...
> ![image](https://user-images.githubusercontent.com/70162741/188012753-223f9dbe-7633-4957-80b4-5ad7fb965c47.png)
> I actually made them the same exact color as the `onion-skinning` icons.

Onion skinning uses a dark color because the icons have thin strokes and transparency. Zoom controls use `#575e75`.

ErrorGamer2000 commented:
Oh, okay. Change it soon.

ErrorGamer2000 commented:
Fixed

Norbiros reviewed (changes requested):
1. When I am dynamic enabling addon I have this error:
```log
Uncaught TypeError: state is undefined [userscript.js:124:7]
```
2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
![Screenshot from 2022-09-06 19-07-19](https://user-images.githubusercontent.com/89837724/188696913-48fcf775-3b7a-4554-90d1-6e77257eaaf6.png)
3. Can we add setting, or can we just change line colour?

ErrorGamer2000 commented:
> 1. When I am dynamic enabling addon I have this error:
> ```log
> Uncaught TypeError: state is undefined [userscript.js:124:7]
> ```
> 2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
> ![Screenshot from 2022-09-06 19-07-19](https://user-images.githubusercontent.com/89837724/188696913-48fcf775-3b7a-4554-90d1-6e77257eaaf6.png)
> 3. Can we add setting, or can we just change line colour?
> 

1. @Norbiros that error is not possible since `userscript.js` does not have a line 124, and no other file that I added uses `state` on line 124.
2. It's a "feature"
3. Sure, have that done in a little while.

ErrorGamer2000 commented:
> 1. When I am dynamic enabling addon I have this error:
> ```log
> Uncaught TypeError: state is undefined [userscript.js:124:7]
> ```
> 2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
> ![Screenshot from 2022-09-06 19-07-19](https://user-images.githubusercontent.com/89837724/188696913-48fcf775-3b7a-4554-90d1-6e77257eaaf6.png)
> 3. Can we add setting, or can we just change line colour?
> 

1. @Norbiros that error is not possible since `userscript.js` does not have a line 124, and no other file that I added uses `state` on line 124.
2. It's a "feature"
3. Sure, have that done in a little while.

Norbiros commented:
LGTM

ErrorGamer2000 commented:
Will fix in an hour or so, but that's minor and causes no issues.

apple502j commented:
@Norbiros just to make sure, did you have default costume editor color enabled when it errored?

ErrorGamer2000 commented:
@apple502j I have it enabled too. And no error.

ErrorGamer2000 commented:
@apple502j and @WorldLanguages please review again, I think I have all of the bugs worked out.

TheColaber commented on addons/paint-snap/state.js:
```diff
@@ -0,0 +1,67 @@
+
+export let guideColor;
+
+if (ls.getItem("sa-paint-snap-snapTo")) {
+  const stored = JSON.parse(ls.getItem("sa-paint-snap-snapTo"));
+  Object.assign(snapTo, stored);
+}
+
+if (ls.getItem("sa-paint-snap-snapFrom")) {
+  const stored = JSON.parse(ls.getItem("sa-paint-snap-snapFrom"));
+  Object.assign(snapFrom, stored);
+}
+
+if (ls.getItem("sa-paint-snap-threshold")) {
+  const stored = JSON.parse(ls.getItem("sa-paint-snap-threshold"));
+  threshold = stored;
+}
+
+export function setSnapTo(point, to) {
+  snapTo[point] = !!to;
+  ls.setItem("sa-paint-snap-snapTo", JSON.stringify(snapTo));
+}
+export function setSnapFrom(point, to) {
+  snapFrom[point] = !!to;
+  ls.setItem("sa-paint-snap-snapFrom", JSON.stringify(snapFrom));
+}
+
+export function setThreshold(thresh) {
+  threshold = thresh;
+  ls.setItem("sa-paint-snap-threshold", JSON.stringify(threshold));
+}
```
This is inconsistent with onion skinning. @WorldLanguages should we add tons of settings like in onion skinning? Or should we have onion skinning do this? Or maybe we add an API to handle storage?

ErrorGamer2000 replied:
I vote api.

Norbiros replied:
API was created (in #2843) but it wasn't merged...

ErrorGamer2000 replied:
Ummm

![image](https://user-images.githubusercontent.com/70162741/190242296-d7a74069-d817-4d0f-bcdc-929b52b5cd6a.png)

mxmou replied:
Related: #1950

ErrorGamer2000 replied:
somewhat, yes

WorldLanguages replied:
I don't think it makes sense to delay the release of this addon until we have a good way to handle cases like this.

I'd say we make this addon handle settings just like onion-skinning for the initial release of this addon.

ErrorGamer2000 commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,36 @@
+{
+  "$schema": "https://raw.githubusercontent.com/ScratchAddons/manifest-schema/dist/schema.json",
```
Ah, yes.

ErrorGamer2000 commented:
I do have a few small tweaks I think I should make, one of which being renaming both settings with "edge centers" to "edge midpoints".

apple502j commented on addon-api/content-script/Tab.js:
```diff
@@ -356,6 +356,8 @@ export default class Tab extends Listenable {
    * assetContextMenuAfterExport - after the export button of asset (sprite, costume, etc)'s context menu
    * assetContextMenuAfterDelete - after the delete button of asset (sprite, costume, etc)'s context menu
    * monitor - after the end of the stage monitor context menu
+   * paintEditorZoomControls - Before the zoom controls in the paint editor
```
```suggestion
   * paintEditorZoomControls - before the zoom controls in the paint editor
```

apple502j commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,97 @@
+  "name": "Costume editor snapping",
+  "description": "Snap objects in the costume editor to bounding boxes and vector nodes.",
+  "tags": ["editor", "costumeEditor", "featured"],
+  "versionAdded": "1.30.0",
```
```suggestion
  "versionAdded": "1.29.0",
```

ErrorGamer2000 reviewed (commented):
done

Secret-chest commented:
The default config should have all corners and centres and no edges IMO.

GarboMuffin commented:
Some nits:

 - I think it would make sense to automatically turn on snapping when someone changes the settings, like what onion skinning does
 - I think it would look better if the black border between sections of the settings was made less dark or removed
 - If all of the "Snap from:" or "Snap to:" options are disabled, I think it would make sense to make the snapping icon look disabled
 - When all of "Snap from:" or "Snap to:" are disabled and someone tries to turn on snapping, perhaps it would make sense to turn on a couple things for them

ErrorGamer2000 commented:
> Some nits:
> 
> * I think it would make sense to automatically turn on snapping when someone changes the settings, like what onion skinning does
> * I think it would look better if the black border between sections of the settings was made less dark or removed
> * If all of the "Snap from:" or "Snap to:" options are disabled, I think it would make sense to make the snapping icon look disabled
> * When all of "Snap from:" or "Snap to:" are disabled and someone tries to turn on snapping, perhaps it would make sense to turn on a couple things for them

All fixed @GarboMuffin

lisa-wolfgang commented:
https://github.com/ScratchAddons/ScratchAddons/issues/5219 was proposed recently. @ErrorGamer2000, would you be interested in implementing this, or would it be best to create a separate PR? Full disclosure, I don't know how much effort this would take.

ErrorGamer2000 commented:
@lisa-wolfgang That is definitely a viable option. Grid mode would have to be a *separate* type of snapping that you can switch to, however, because the two modes would not work very well together. However, I think it would be best to wait until #5226 is either approved or rejected before making the decision or not, because that would end up with a **very** large number of settings for this addon.

mxmou commented:
> Grid mode would have to be a _separate_ type of snapping that you can switch to, however, because the two modes would not work very well together.

Does this mean that enabling snap to grid would disable snapping to all other objects?

ErrorGamer2000 commented:
@mxmou yes

mxmou commented:
> @mxmou yes

Is there a technical reason for that? I don't think users would expect this restriction - Inkscape, for example, doesn't have it.

ErrorGamer2000 commented:
> > @mxmou yes
> 
> Is there a technical reason for that? I don't think users would expect this restriction - Inkscape, for example, doesn't have it.

True, but the paint area is so small in scratch, the two would just stumble over each other.

lisa-wolfgang commented:
I think it should be fine for the majority of cases, especially with large grid sizes, but what I do in Inkscape in that situation is hide the grid, snap, then show the grid again. If I'm remembering correctly, you add a grid snapping toggle in the editor, so that method should work fine with this addon as well.

ErrorGamer2000 commented:
> I think it should be fine for the majority of cases, especially with large grid sizes, but what I do in Inkscape in that situation is hide the grid, snap, then show the grid again. If I'm remembering correctly, you add a grid snapping toggle in the editor, so that method should work fine with this addon as well.

Hmm ok then

WorldLanguages commented on addons/paint-snap/addon.json:
```diff
@@ -0,0 +1,110 @@
+{
+  "name": "Costume editor snapping",
+  "description": "Snap objects in the costume editor to bounding boxes and vector nodes.",
+  "tags": ["editor", "costumeEditor", "featured"],
```
Name, description and tags LGTM

WorldLanguages commented:
@ErrorGamer2000 Any blockers? Maybe we could ask GarboMuffin for an explicit review?

ErrorGamer2000 commented:
No blockers and I think Garbo already did @WorldLanguages

ErrorGamer2000 commented:
Ok then @GarboMuffin pls check

WorldLanguages commented:
This has 2 approvals, so it can be merged (I'll take a quick look first). We should try to find bugs until the translation process ends in a few days.
Next time, #5384 will be useful to avoid PRs like this going stale for 3 months.

WorldLanguages commented:
We could also decide to add the "beta" tag anytime before actual release. But strings, such as the addon name and description, can't change after today.

RedGuy12 commented:
![image](https://user-images.githubusercontent.com/61329810/210796064-1489bab4-96c8-4fd6-8fa8-76bd676cd624.png)

CST1229 commented:
I think the snapping GUI should be visual, something like this:
![image](https://user-images.githubusercontent.com/68464103/210803695-3906c3e7-cffb-4376-a32b-2e97a79733a1.png)

WorldLanguages commented:
Re: Jazza (screenshot posted by @RedGuy12)

1. Snapping can already be toggled universally for the session through the magnet icon.
2. The find bar going in front is a bug that deserves its own dedicated issue.
3. The menu not hiding when clicking outside is not a new bug, that's how "onion skinning" menu has worked for years. Also needs a dedicated issue.

WorldLanguages commented:
> I think the snapping GUI should be visual, something like this:

@CST1229 That would be helpful to understand what each specific setting does. But it could get annoying.

ErrorGamer2000 commented:
> I think the snapping GUI should be visual, something like this: ![image](https://user-images.githubusercontent.com/68464103/210803695-3906c3e7-cffb-4376-a32b-2e97a79733a1.png)

@CST1229 I honestly considered this. But it would be both difficult to make, and difficult for users to understand.
